File: programmers/Brute-Force/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

bool cmp(vector<int> a, vector<int> b) {
    if(a[0] < b[0]) return true;
    if(a[0] == b[0] && a[1] < b[1]) return true;
    return false;
}

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    sort(dungeons.begin(), dungeons.end(), cmp);
    do {
        int cur = k, cnt = 0;
        for(auto dungeon : dungeons) {
            if(cur < dungeon[0]) break;
            cur -= dungeon[1];
            cnt++;
        }
        answer = max(answer, cnt);
    } while (next_permutation(dungeons.begin(), dungeons.end()));
    
    return answer;
}

File: 프로그래머스/lv2/87946. 피로도/피로도.java
class Solution {
    int[][] dungeons;
    int answer = 0;
    public int solution(int k, int[][] dungeons) {
        this.dungeons = dungeons;
        boolean[] visit = new boolean[dungeons.length];
        dfs(k, visit, 0);
        return answer;
    }
    
    void dfs(int k, boolean[] visit, int cnt) {
        answer = Math.max(answer, cnt);
        for (int i=0; i<visit.length; i++) {
            if (visit[i] || k < dungeons[i][0]) continue;
            visit[i] = true;
            dfs(k-dungeons[i][1], visit, cnt+1);
            visit[i] = false;
        }
    }
}